File: ts/tests/DeviceLink.test.ts
import { generateSecretKey, getPublicKey } from "nostr-tools"
import { describe, expect, it } from "vitest"

import {
  createDeviceLinkRequest,
  deterministicLinkInviteForDeviceLinkRequest,
  parseCompactDeviceLinkRequest,
} from "../src/DeviceLink"

describe("DeviceLink", () => {
  it("encodes scan-to-approve requests as one device pubkey plus one request secret", () => {
    const deviceAppKeySecretKey = generateSecretKey()
    const requestSecretKey = generateSecretKey()
    const local = createDeviceLinkRequest({
      deviceAppKeySecretKey,
      requestSecretKey,
      requestedAt: 41,
      label: "Iris Chat Web",
    })

    expect(local.code).toMatch(/^[0-9a-f]{64}\.[0-9a-f]{64}$/)
    expect(local.code).not.toContain("https://chat.iris.to")
    expect(local.code).not.toContain("nostrconnect:")
    expect(local.code).not.toContain("relay=")
    expect(local.code.length).toBe(129)
    expect(local.request.deviceAppKeyPubkey).toBe(getPublicKey(deviceAppKeySecretKey))
    expect(local.request.requestPubkey).toBe(getPublicKey(requestSecretKey))
    expect(local.request.requestSecret).toBe(toHex(requestSecretKey))

    expect(local.code.split(".")).toEqual([
      local.request.deviceAppKeyPubkey,
      toHex(requestSecretKey),
    ])

    const parsed = parseCompactDeviceLinkRequest(local.code)
    expect(parsed?.requestPubkey).toBe(local.request.requestPubkey)
    expect(parsed?.deviceAppKeyPubkey).toBe(local.request.deviceAppKeyPubkey)
    expect(parsed?.requestSecret).toBe(local.request.requestSecret)
    expect(parsed?.label).toBeUndefined()
  })

  it("rejects malformed approval codes", () => {
    expect(parseCompactDeviceLinkRequest("https://chat.iris.to/")).toBeNull()
    expect(parseCompactDeviceLinkRequest("not-a-compact-link-code")).toBeNull()
    expect(parseCompactDeviceLinkRequest(`${"a".repeat(64)}.${"b".repeat(64)}.extra`)).toBeNull()
  })

  it("derives deterministic NDR link invites from the compact secret", () => {
    const requestSecret =
      "0100000017000000c8010000d21e000000000000000000000000000000000000"
    const deviceAppKeyPubkey = "e".repeat(64)
    const request = {
      requestPubkey: getPublicKey(hexToBytes(requestSecret)),
      deviceAppKeyPubkey,
      requestSecret,
      requestedAt: 77,
    }

    const invite = deterministicLinkInviteForDeviceLinkRequest(request)
    const repeated = deterministicLinkInviteForDeviceLinkRequest({
      ...request,
      requestedAt: 88,
    })

    expect(toHex(invite.inviterEphemeralPrivateKey!)).toMatch(/^be3f1cca6354c294/)
    expect(invite.inviterEphemeralPublicKey).toBe(repeated.inviterEphemeralPublicKey)
    expect(invite.sharedSecret).toBe(repeated.sharedSecret)
    expect(invite.inviter).toBe(deviceAppKeyPubkey)
    expect(invite.deviceId).toBe(deviceAppKeyPubkey)
    expect(invite.maxUses).toBe(1)
    expect(invite.createdAt).toBe(0)
    expect(invite.purpose).toBe("link")
  })
})

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("")
}

function hexToBytes(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2)
  for (let index = 0; index < bytes.length; index += 1) {
    bytes[index] = Number.parseInt(hex.slice(index * 2, index * 2 + 2), 16)
  }
  return bytes
}

File: rust/crates/nostr-double-ratchet/tests/device_link_test.rs
use nostr::Keys;
use nostr_double_ratchet::{
    deterministic_link_invite_for_device_link_request, encode_compact_device_link_request,
    parse_compact_device_link_request,
};

#[test]
fn compact_device_link_request_round_trips() {
    let device = Keys::generate();
    let request = Keys::generate();
    let code = encode_compact_device_link_request(
        device.public_key(),
        &request.secret_key().to_secret_hex(),
    )
    .expect("encode compact request");

    assert_eq!(code.len(), 129);
    let parsed = parse_compact_device_link_request(&code).expect("parse compact request");
    assert_eq!(parsed.device_app_key_pubkey, device.public_key());
    assert_eq!(parsed.request_pubkey, request.public_key());
    assert_eq!(parsed.request_secret, request.secret_key().to_secret_hex());
}

#[test]
fn compact_device_link_request_rejects_malformed_inputs() {
    assert!(parse_compact_device_link_request("").is_err());
    assert!(parse_compact_device_link_request("npub1plainvalue").is_err());
    assert!(parse_compact_device_link_request("https://example.com").is_err());
    assert!(parse_compact_device_link_request(&format!(
        "{}.{}.extra",
        "1".repeat(64),
        "1".repeat(64)
    ))
    .is_err());
}

#[test]
fn deterministic_device_link_invite_matches_typescript_vector() {
    let request_secret = "0100000017000000c8010000d21e000000000000000000000000000000000000";
    let device_pubkey = "e".repeat(64);
    let request = parse_compact_device_link_request(&format!("{device_pubkey}.{request_secret}"))
        .expect("parse vector");

    let invite =
        deterministic_link_invite_for_device_link_request(&request).expect("create invite");
    let repeated =
        deterministic_link_invite_for_device_link_request(&request).expect("repeat invite");

    assert!(
        hex::encode(invite.inviter_ephemeral_private_key.unwrap()).starts_with("be3f1cca6354c294")
    );
    assert_eq!(
        invite.inviter_ephemeral_public_key,
        repeated.inviter_ephemeral_public_key
    );
    assert_eq!(invite.shared_secret, repeated.shared_secret);
    assert_eq!(invite.inviter.to_hex(), device_pubkey);
    assert_eq!(invite.device_id.as_deref(), Some(device_pubkey.as_str()));
    assert_eq!(invite.max_uses, Some(1));
    assert_eq!(invite.created_at.get(), 0);
    assert_eq!(invite.purpose.as_deref(), Some("link"));
}
